File: backend/app/schemas/project.py
from typing import List, Optional, Literal
from pydantic import BaseModel, constr
from datetime import datetime
from .task import Task  # Import Task schema


class ProjectBase(BaseModel):
    name: constr(min_length=1, max_length=200)
    description: constr(max_length=5000)
    start_date: datetime
    deadline: datetime


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(ProjectBase):
    status: Optional[str] = None
    name: Optional[constr(min_length=1, max_length=200)] = None
    description: Optional[constr(max_length=5000)] = None
    start_date: Optional[datetime] = None
    deadline: Optional[datetime] = None


class ProjectMemberBase(BaseModel):
    user_id: str
    role: Literal['project_member', 'project_manager'] = 'project_member'


class ProjectMember(ProjectMemberBase):
    id: str
    project_id: str
    created_at: datetime


class Project(ProjectBase):
    id: str
    status: str
    created_at: datetime
    updated_at: datetime
    is_active: bool
    tasks: List[Task] = []  # Add tasks field with default empty list
    members: List[ProjectMember] = []  # Add members field with default empty list

    class Config:
        from_attributes = True

File: frontend/src/lib/types/project.ts
export interface Project {
    id: string;
    name: string;
    description: string;
    status: string;
    start_date: string;
    deadline: string;
    tasks?: Task[];
    members?: ProjectMember[];
}

export interface ProjectCreate {
    name: string;
    description: string;
    start_date: string;
    deadline: string;
}

export interface ProjectMember {
    id: string;
    user_id: string;
    project_id: string;
    role: string;
    name?: string;
    email?: string;
}

export interface Task {
    id: string;
    name: string;
    description: string;
    project_id: string;
    parent_task_id?: string;
    start_date: string;
    deadline: string;
    assigned_to_id?: string;
    created_by_id: string;
    status: 'pending' | 'in_progress' | 'completed';
    created_at: string;
    updated_at: string;
    subtasks?: Task[];
}

export interface TaskCreate {
    name: string;
    description: string;
    project_id: string;
    start_date: string;
    deadline: string;
    assigned_to_id?: string | null;
}

export interface GanttTask {
    id: string;
    name: string;
    start: string;
    end: string;
    progress: number;
    dependencies?: string[];
}

export interface Comment {
    id: string;
    task_id: string;
    user_id: string;
    content: string;
    created_at: string;
    updated_at: string;
}

export interface CommentCreate {
    task_id: string;
    content: string;
} 